### PR TITLE
Verify if snap name is registered before requesting new snap

### PR DESCRIPTION
**Pull request by bartaz**
screen shot 2017-01-05 at 14 25 31

### CODE REVIEW

File: environments/development.env [hunk 1]
```diff
@@ -6,6 +6,7 @@ NODE_ENV=development
 UBUNTU_SSO_URL=https://login.ubuntu.com
 OPENID_VERIFY_URL=http://localhost:8000/login/verify
 LP_API_URL=https://api.launchpad.net
+STORE_MACAROON_API_URL=https://myapps.developer.ubuntu.com/dev/api
```
Comment (bartaz) [inline environments/development.env]: As we are talking to production LP only, I guess it's fine to talk to production store to check name registration.

File: src/server/handlers/launchpad.js [hunk 2]
```diff
@@ -210,6 +218,47 @@ const getSnapcraftYaml = (owner, name, token) => {
...
+  }).then((response) => response.json().then((json) => {
+    if (response.status === 200 && json.macaroon) {
+      return name;
+    } else {
+      throw new PreparedError(400, RESPONSE_SNAP_NAME_NOT_REGISTERED);
+    }
+  }));
+};
+
+const requestNewSnap = (name, repositoryUrl) => {
```
Comment (bartaz) [inline src/server/handlers/launchpad.js]: Extracted it from body of `newSnap` for easier promise chaining.

File: test/routes/src/server/routes/t_launchpad.js [hunk 3]
```diff
@@ -74,113 +74,163 @@ describe('The Launchpad API endpoint', () => {
       });
     });
 
-    context('when snap does not exist', () => {
-      const caveatId = 'dummy caveat';
+    context('when repo contains valid snapcraft.yaml', () => {
```
Comment (bartaz) [inline test/routes/src/server/routes/t_launchpad.js]: Quite a lot of changes here due to moving tests around.

I moved some tests that shared same mocked requests (valid snapcraft.yaml, name registered) into common nested contexts to avoid duplicating same calls to `nock`.

File: test/routes/src/server/routes/t_launchpad.js [hunk 4]
```diff
@@ -74,113 +74,163 @@ describe('The Launchpad API endpoint', () => {
...
+      const snapName = 'dummy-test-snap';
 
       beforeEach(() => {
         nock(conf.get('GITHUB_API_ENDPOINT'))
           .get('/repos/anaccount/arepo')
           .reply(200, { permissions: { admin: true } });
         nock(conf.get('GITHUB_API_ENDPOINT'))
           .get('/repos/anaccount/arepo/contents/snapcraft.yaml')
-          .reply(200, 'name: test-snap\n');
```
Comment (bartaz) [inline test/routes/src/server/routes/t_launchpad.js]: Fun fact: the tests were actually passing with `test-snap` (even though macaroon requests were not mocked), because `test-snap` seems to be actually a real registered name in public store, so when tests called real API they got valid macaroon response.

File: docs/environment-variables.md [hunk 5]
```diff
@@ -67,6 +67,15 @@ The OAuth access token, generated by Launchpad.
...
 The OAuth token secret, generated by Launchpad.
 
+
+## Ubuntu Store settings
+
+### STORE\_MACAROON\_API\_URL
+
+- **Example**: `https://myapps.developer.ubuntu.com/dev/api`
+
+The URL of the Ubuntu Store Macaroon API service.
```
Comment (cjwatson) [inline docs/environment-variables.md]: I'd call this just `STORE_API_URL` and "the Ubuntu Store API service", I think; there's more than just the macaroon API under there, and we might want some other bit of it in future.

File: environments/development.env (hunk 1 above)
Comment (cjwatson) [inline environments/development.env]: Indeed - you have no choice.

File: src/server/handlers/launchpad.js [hunk 6]
```diff
@@ -210,6 +218,47 @@ const getSnapcraftYaml = (owner, name, token) => {
     });
 };
 
+const verifySnapNameRegistered = (name) => {
+  return fetch(`${conf.get('STORE_MACAROON_API_URL')}/acl/`, {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
```
Comment (cjwatson) [inline src/server/handlers/launchpad.js]: I can't remember if `fetch` does this by itself, but we should make sure to pass `Accept: application/json` as well, even if the store doesn't enforce that today.

File: test/routes/src/server/routes/t_launchpad.js (hunk 4 above)
Comment (cjwatson) [inline test/routes/src/server/routes/t_launchpad.js]: This can wait until later, but I wonder if we should have some global test setup that puts some kind of catch-all request mock in place so that requests don't slip through to production services if we forget something.

File: test/routes/src/server/routes/t_launchpad.js [hunk 7]
```diff
@@ -74,113 +74,163 @@ describe('The Launchpad API endpoint', () => {
...
-          .post('/launchpad/snaps')
-          .send({ repository_url: 'https://github.com/anaccount/arepo' })
-          .expect(hasStatus('success'))
-          .end(done);
-      });
+        beforeEach(() => {
+          nock(conf.get('STORE_MACAROON_API_URL'))
+            .post('/acl/', {
+              'packages': [{ 'name': snapName, 'series': '16' }],
+              'permissions':['package_upload']
```
Comment (cjwatson) [inline test/routes/src/server/routes/t_launchpad.js]: Space after colon?  (Similarly below.)

File: src/server/handlers/launchpad.js [hunk 8]
```diff
@@ -83,6 +83,14 @@ const RESPONSE_SNAPCRAFT_YAML_NO_NAME = {
   }
 };
 
+const RESPONSE_SNAP_NAME_NOT_REGISTERED = {
+  status: 'error',
+  payload: {
+    code: 'snap-name-not-registered',
+    message: 'Snap name is not registered in the store'
```
Comment (evandandrea) [inline src/server/handlers/launchpad.js]: It could wait until later, but would it be quick to change this to "The name provided in the snapcraft.yaml ($name) is not registered in the store. Please register it \<a href="https://myapps.developer.ubuntu.com/dev/click-apps/register-name/?name=$name">here</a> before trying again."

Comment (cjwatson) [inline src/server/handlers/launchpad.js]: I'd suggest that we should do this by handling that payload code specially in the frontend component, rather than by injecting HTML from the server-side handler.

Comment (bartaz) [inline src/server/handlers/launchpad.js]: Yes, it should be handled in views, as HTML in such messages would be escaped by React.

Sure, ideally we want actionable error message like this, but for MVP demo this should be enough. And it would be great to have some designs if we start to look into details like this.

Comment (cjwatson) [inline src/server/handlers/launchpad.js]: I'm going to do this in a separate PR, rather than adding it to Bartek's.
